[Ian Hong] iP

ianfromdover opened the pull request:
# Lily
> “Why is your list so full of nonsense” – Lily

Lily is a CLI-based task manager. It attemps to embody the personality of a character in Duolingo of the same name. She is a disinterested teen with purple hair.


![Lily from Duolingo](https://c.tenor.com/8WRjkoq7QF0AAAAC/duolingo-lily.gif)

- On an emotional scale of 1-10, **Lily is at a -4**
- Lily's perpetually ~~enthusiastic~~unamused facial expression has one clear message: stop bothering me
- If you check her Spotify, she *might* have [Girl in Red](https://open.spotify.com/artist/3uwAm6vQy7kWPS2bciKWx9?si=J5xVJ9WNRwGRc0Xv8WBHjg) in her playlist. Who needs therapy when you can just dye your hair? 🥀

## Features
Example of what you can track with Lily:
- [ ] Todos, without a date
- [ ] Events which has a date (at: Evening of 2 Feb 2022)
- [x] Deadlines which has a due date (by: Wednesday 4 Feb 2022)
---
This is what you should see when launching her each time:
```

       ▼＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝▼ 
       ██╗     ██╗██╗     ██╗   ██╗
       ██║     ██║██║     ╚██╗ ██╔╝
       ██║     ██║██║      ╚████╔╝ 
       ██║     ██║██║       ╚██╔╝  
       ███████╗██║███████╗   ██║   
       ╚══════╝╚═╝╚══════╝   ╚═╝   
       hey.
       ▼＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝▼
```
which corresponds to
```java
public void showWelcome(TaskList tl) {
    String welcomeMessage = logo;
    if (!couldLoadSaveData) {
        welcomeMessage += INDENT + "hey." + LS;
    }
    prettyPrint(welcomeMessage);
}
```

## Installation
1. Download Lily from the [Releases page](https://github.com/ian-from-dover/ip/releases)
2. Launch the .jar file by double clicking on it

Commit:
Add Gradle support

ianfromdover committed:
Add Personality Plans

ianfromdover committed:
Changed name of Duke to Lily

ianfromdover committed:
Fixed echo system, Added Indents

ianfromdover committed:
Made Lily be able to Add and List Todos

ianfromdover committed:
Implemented Lily Marking and Unmarking Tasks without OOP

ianfromdover committed:
Created Task Object, Added PrettyPrint

ianfromdover committed:
Implemented LinkedList for Tasks, Added Events, Todos and Deadlines

ianfromdover committed:
Prepare to create new Task Objects

ianfromdover committed:
Add Inheritance for Todo, Deadline and Event

ianfromdover committed:
Test New Remote

ianfromdover committed:
Push to new origin

ianfromdover committed:
Automate Output Testing with runtest.sh

ianfromdover committed:
Made LilyException Class but need to test throwing it, moved string parting into classes

ianfromdover committed:
Test Exceptions Create Lily Child Exceptions

ianfromdover committed:
Implemented Delete and Remove functionality

ianfromdover committed:
Added Enums

ianfromdover committed:
Add File Writing Functionality

ianfromdover committed:
Parse Dates as LocalDates

ianfromdover committed:
Add Date and Save function to master

ianfromdover committed:
Package Tasks and Add Javadocs

ianfromdover committed:
Break Lily class into smaller classes (buggy)

ianfromdover committed:
Encapsulate TaskList, Storage and break Lily down

ianfromdover committed:
Break into further tasks for Parser

ianfromdover committed:
Make OOP restructure work

Separate files for Ui, Parsing, the TaskList and Storage have been
created, but they do not work. the Lily class also contains all of the old reference code.

Users who wish to use this app need it to work when they download it.

Let's makes the application work again by fixing all the bugs and
placing each code block into its respective class.

This conforms to the OOP programming paradigm.

Assume all commits since the previous tag do not work.

ianfromdover committed:
Standardise package and imports

Let's order each statement by package exceptions, package classes, data structures, and then IO.

ianfromdover committed:
Update Readme, Add Gradle, Write JavaDocs

ianfromdover committed:
Adhere to coding standard

ianfromdover committed:
Update version month

ianfromdover committed:
Align with coding standards

ianfromdover committed:
Add find functionality

ianfromdover committed:
Merge branch 'branch-A-JavaDoc'

Fix extra javadocs which were not previously written

ianfromdover committed:
Fix conflicts when merging CodingStandard to JavaDocs

ianfromdover committed:
Merge branch 'branch-Level-9'

Add find functionality to master

Dineshraj555 commented on src/main/java/lily/TaskList.java:
```diff
@@ -0,0 +1,147 @@
+     * @throws IndexOutOfBoundsException If the index exceeds the list.
+     * @throws LilyException If user hasn't done the task yet.
+     */
+    public Task unmark(int idx) throws IndexOutOfBoundsException, LilyException {
```
I believe camelCase should be used here. unMark.

ianfromdover replied:
`unmark` is a real word on its own so it should be okay to leave it as it is :)

Dineshraj555 commented on src/main/java/lily/task/Task.java:
```diff
@@ -0,0 +1,78 @@
+     * 
+     * @throws LilyException If user hasn't done the task yet.
+     */
+    public void unmark() throws LilyException {
```
Same as above. I believe camelCase should be used here. unMark.

Dineshraj555 reviewed (commented):
LGTM. Well written and easy to follow through your code. 2 minor coding standard violations.

tehkokhoe reviewed (commented):
A fun, easy code to read, I enjoyed it. I couldn't find any mistakes either, good job! :)

viki0526 commented on src/main/java/lily/Parser.java:
```diff
@@ -0,0 +1,165 @@
+                sentence = ui.readCommand();
+                parsedSentence = sentence.split(" ");
+                command = parsedSentence[0];
+                switch (command) {
+                case "bye":
+                    st.save(tasks);
+                    break userInteracting;
+    
+                case "list":
+                    ui.showList(tasks);
+                    break;
+    
+                case "mark":
+                    int addIdx = Integer.parseInt(parsedSentence[1]); // base 1
+                    ui.showMarked(tasks.mark(addIdx - 1), addIdx);
+                    break;
+    
+                case "unmark":
+                    int delIdx = Integer.parseInt(parsedSentence[1]);
+                    ui.showUnmarked(tasks.unmark(delIdx - 1), delIdx);
+                    break;
+    
+                case "todo":
+                    ui.showTaskAdded(tasks.addTodo(findTodoDescStart(sentence)), tasks.getSize());
```
Nice, clean use of switch-case with appropriate spacing to parse commands

ianfromdover replied:
Thanks!

viki0526 commented on src/main/java/lily/Ui.java:
```diff
@@ -0,0 +1,241 @@
+                LS + "    ╚══════╝╚═╝╚══════╝   ╚═╝   " + LS + LS;
+        String welcomeMessage = logo;
+
+        if (couldLoad) {
+            welcomeMessage += INDENT + "sup, welcome back" + LS
+                    + INDENT + "here's your list from the last time" + LS
+                    + LS
+                    + listTasks(tl.getContents()) + LS + LS
+                    + INDENT + "why is it so full of nonsense";
+        } else {
+            showError(INDENT + "oh bother, a new one");
+            welcomeMessage += INDENT + "hey." + LS
+                    + INDENT + "i don't really wanna track your tasks," + LS
+                    + INDENT + "but i guess i have no choice (っ◞‸◟c)" + LS
+                    + INDENT + "need help with something?" + LS
+                    + LS
+                    + INDENT + "Things you can type" + LS
+                    + getCommands();
+        }
```
Good use of indentation when using the '+' operator

ianfromdover replied:
Thanks!

viki0526 commented on src/main/java/lily/Ui.java:
```diff
@@ -0,0 +1,241 @@
+     * 
+     * @return The user's input as a string.
+     */
+    public String readCommand() {
+        return userInput.nextLine();
+    }
+
+    /**
+     * Prints the exit message and closes the scanner.
+     */
+    public void closeUi() {
+        prettyPrint("finally. what took you so long? (´-ω-`)" + LS + LS
+                + INDENT + "oh, if your list had stuff, i've saved it" + LS
+                + INDENT + "i'll bring it up when you bug me again. bye");
+        userInput.close();
+    }
+
+    /**
+     * Displays the items in the TaskList to the user.
+     * 
+     * @param tl The TaskList to be shown.
+     */
+    public void showList(TaskList tl) {
+        if (tl.isEmpty()) {
+            showError("there's nothing in the list bro");
+        } else {
+            prettyPrint("you told me you had to" + LS + listTasks(tl.getContents()));
+        }
+    }
```
Good use of camelCasing and descriptive verbs to name methods

ianfromdover replied:
Thanks!

viki0526 reviewed (commented):
Very neat and readable code and follows coding standards very closely. LGTM.

ianfromdover committed:
Remove images, align README with pull req desc

ianfromdover committed:
Move to IntelliJ, Create Jar, Fix data writing

Export Jar file, the jar file does not save the list on exiting.

Users downloading the released jar file need to save their progress
after exiting.

Let's,
* Fix the save bug and allow users to save their progress by changing
the output file to './data' instead of '/data'
* Export a JAR file for easy distribution
* Add more intuitive ways to mark a task as done such as 'do' and 'done'

The first fix is effective in removing the error of being unable to
save. Users usually refer to completing tasks as 'do' and 'done' instead
of 'mark'.

ianfromdover committed:
Prepare for merge with gradle support

ianfromdover committed:
Merge remote-tracking branch 'new-origin/add-gradle-support'

ianfromdover committed:
Add JUnit Tests with Gradle

Testing is not automated.

Automating tests reduces manual work needed to test when refactoring.

Let's add unit tests so that we know whether the code changes still
keep the code functional.

ianfromdover committed:
Move to GUI, add Lily images

The application only works on command line.

Most consumers do not know how to run such applications.

Let's make the application interactable through a GUI.

ianfromdover committed:
Add Varargs to Todo, Update user guide

ianfromdover committed:
Parser: Add assertions. LilyException: Add more exceptions

ianfromdover committed:
(in-process) Improve code quality

ianfromdover committed:
Fix code quality, abstract out common methods

ianfromdover committed:
Merge pull request #1 from ian-from-dover/branch-A-CodeQuality

Improve code quality

ianfromdover committed:
Merge Assertions branch into master

ianfromdover committed:
Add CI

ianfromdover committed:
Add Job task

ianfromdover committed:
Add Job task file

ianfromdover committed:
Set theme jekyll-theme-cayman

ianfromdover committed:
Fix JavaDocs to follow coding standards

ianfromdover committed:
Merge website docs to local coding standards version

ianfromdover committed:
Prepare for Ui design

ianfromdover committed:
Polish GUI, add stickers, add welc, exit msgs

GUI was copied from the tutorial and was not optimised for Lily's
personality.

Having a GUI which suits the theme of the app and the tone of the
character makes the experience more immersive.

Let's,
* select colours which reflect the duolingo colour scheme
* choose a similar font to that reflected in duolingo's style guide
* reference classmates's scripts for fxml inspiration
* separate dialogbox classes
* add stickers to convey emotions
* add welcome and exit messages
* add pauses between Lily responding to the user giving a more
  humanlike chatting experience

ianfromdover committed:
Update ui screenshot

ianfromdover committed:
Polish READMEs, improve code quality

Let's
* Add more pictures to the READMEs
* Update the user guide to cover all the features
* Remove redundant code
* Write JavaDocs for GUI methods

ianfromdover committed:
Fix Code Quality Issues

ianfromdover committed:
remove unused imports

ianfromdover committed:
change parser
